### PR TITLE
add Large conifers

### PR DESCRIPTION
**Pull request by mBornand**
Adds the large conifers trees for the mega taiga biomes and mega spruce taiga.

### CODE REVIEW

**Comment by mBornand** [comment]
In https://github.com/cuberite/cuberite/issues/4031#issuecomment-330148583 @madmaxoft said that `constexpr` should not be used. Is this still up to date?

**Comment by mBornand** [comment]
I think this will fix #376
I just didn't see this Issue before.

**Comment by tigerw** [comment]
I think we need some screenshots :)

**Comment by mBornand** [comment]
And the Spruce only variant.

File: src/Generating/Trees.cpp [hunk 1]
```diff
@@ -748,24 +831,58 @@ void GetTallBirchTreeImage(Vector3i a_BlockPos, cNoise & a_Noise, int a_Seq, sSe
 
 
 
-void GetConiferTreeImage(Vector3i a_BlockPos, cNoise & a_Noise, int a_Seq, sSetBlockVector & a_LogBlocks, sSetBlockVector & a_OtherBlocks)
+void GetConiferTreeImage(Vector3i a_BlockPos, cNoise & a_Noise, int a_Seq, sSetBlockVector & a_LogBlocks, sSetBlockVector & a_OtherBlocks, bool a_Large)
 {
 	// Half chance for a spruce, half for a pine:
```
Comment (bearbin) [inline src/Generating/Trees.cpp:836]: This comment needs an update to specify that there is more chance of a large pine than a large spruce.

Comment (mBornand) [reply src/Generating/Trees.cpp]: Now it is changed.

**Comment by bearbin** [comment]
This needs a review from @madmaxoft, but it looks fine to me.

**Comment by Titaniumtown** [comment]
bump, is this going to get pushed?

File: src/Generating/Trees.cpp [hunk 2]
```diff
@@ -862,7 +979,112 @@ void GetSpruceTreeImage(Vector3i a_BlockPos, cNoise & a_Noise, int a_Seq, sSetBl
...
+	{
+		unsigned int Val = (a_Noise.IntNoise3DInt(a_BlockPos.addedXZ(32 * a_Seq, 32 * i)) / 23) % 8;
+		if ((Val < 4) && RingRadius <= 3)
+		{
+			RingRadius++;
+		}
+		else if ((Val >= 7) && (RingRadius <= 0))
+		{
+			RingRadius--;
+		}
```
Comment (madmaxoft) [inline src/Generating/Trees.cpp:1045]: I suspect this condition is wrong, you never want RingRadius to be negative, yet the condition checks `<= 0`

File: src/Generating/Trees.cpp [hunk 3]
```diff
@@ -862,7 +979,112 @@ void GetSpruceTreeImage(Vector3i a_BlockPos, cNoise & a_Noise, int a_Seq, sSetBl
...
+
+
+const int MIN_LARGE_SPRUCE_TREE_RINGS = 3;
+const int MAX_LARGE_SPRUCE_TREE_RINGS = 11;
+void GetLargeSpruceTreeImage(Vector3i a_BlockPos, cNoise & a_Noise, int a_Seq, sSetBlockVector & a_LogBlocks, sSetBlockVector & a_OtherBlocks)
+{
+	int Height = 20 + (a_Noise.IntNoise3DInt(a_BlockPos.addedXZ(32 * a_Seq, 32 * a_Seq)) / 11) % 12;
+	int LeavesRingCount =
+		MIN_LARGE_SPRUCE_TREE_RINGS +
+		(a_Noise.IntNoise3DInt(a_BlockPos.addedXZ(32 * a_Seq, 32 * a_Seq)) / 13) % (1 + MAX_LARGE_SPRUCE_TREE_RINGS - MIN_LARGE_SPRUCE_TREE_RINGS);
```
Comment (madmaxoft) [inline src/Generating/Trees.cpp:1001]: These two random numbers are actually the same, so the `Height` and `LeavesRingCount` are somewhat correlated. Is that intended? If not, change the multipliers to `a_Seq`

**Comment by madmaxoft** [comment]
Normally I don't bother reviewing PRs that fail to build, expecting the authors to fix them or report that the CI is broken.

**Comment by bearbin** [comment]
Merged in https://github.com/cuberite/cuberite/commit/d6c5ac40a6e3495f149fd689edff950eb9ba13c9

**Pull request closed without merging**